UM/Qt/Bindings/MainWindow.py

```diff
@@ -63,14 +63,22 @@ def __init__(self, parent = None):
                                   int(self._preferences.getValue("general/window_width")),
                                   int(self._preferences.getValue("general/window_height")))
 
-        if not self._preferences.getValue("general/restore_window_geometry"):
+        if (not self._preferences.getValue("general/restore_window_geometry") or
+                self._preferences.getValue("general/last_shutdown_severity") == Application.ShutdownSeverity.HardCrash):
             # Ignore whatever the preferences said.
-            Logger.log("i", "Not restoring window geometry from preferences because 'restore_window_geometry' is false")
+            Logger.log("i", "Window geometry not restored from preferences. "
+                            "See 'general/restore_window_geometry' and/or 'general/last_shutdown_severity' for cause.")
             restored_geometry = QRect(self.DEFAULT_WINDOW_LEFT,
                                       self.DEFAULT_WINDOW_TOP,
                                       self.DEFAULT_WINDOW_WIDTH,
                                       self.DEFAULT_WINDOW_HEIGHT)
 
+        # Now set the 'last shutdown normal' to False, then, when the app closes normally, set it back to True.
+        self._preferences.setValue("general/last_shutdown_severity", Application.ShutdownSeverity.HardCrash)
+        Application.getInstance().applicationShuttingDown.connect(
+            lambda: self._preferences.setValue("general/last_shutdown_severity",
+                                               Application.ShutdownSeverity.NormalShutdown))
```

Comment:
From what I could test, this doesn't work:
1. Calling a lambda as a slot doesn't seem to work at all
2. Even if you put the content in a member function, the `applicationShuttingDown` signal is called after the `Application.savePreferences()` is processed, so the `NormalShutdown` state is never saved to file.

Reply:
Ah right, it works for `pyqtSignal`, not for `Signal`.

As for the second point, you're (again) correct. Instead of relying on signals, it may be better to directly call the main window, as it's a local variable there (in the class that sends the signal) anyway.

Reply:
What's the point of using a custom `Signal` mechanism ?